src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java

```diff
@@ -0,0 +1,399 @@
+                if (this == o) return true;
+                if (o == null || getClass() != o.getClass()) return false;
+                NodePreds nodePreds = (NodePreds) o;
+                return Arrays.equals(preds, nodePreds.preds) && opcode == nodePreds.opcode && mode.equals(nodePreds.mode);
+            }
+
+            @Override
+            public int hashCode() {
+                int result = Objects.hash(opcode);
+                result = 31 * result + Arrays.hashCode(preds);
+                return result;
+            }
+        }
+
+        @Override
+        public void defaultVisit(Node node) {
+            // change nothing
+        }
+
+        private void visitPreds(Node node, Node[] preds, binding_irnode.ir_opcode opcode) {
+            NodePreds np = new NodePreds(preds, opcode, node.getMode());
+            if (nodeCache.containsKey(np)) {
+                Node n = nodeCache.get(np);
+                switch (opcode) {
+                    case iro_Conv -> {
+                        Conv orig = (Conv) node;
+                        Conv rep = (Conv) n;
+                        if (orig.getMode().equals(rep.getMode())) {
+                            nodeValues.put(node, n);
+                            addOuts(node);
+                        }
+                    }
+                    default -> {
+                        // can exchange node with already found in list
+                        nodeValues.put(node, n);
+                        // enqueue outs
+                        addOuts(node);
+                    }
+                }
+            } else {
+                nodeCache.put(np, node);
+            }
+        }
+
+        private void addOuts(Node n) {
+            for (var out : BackEdges.getOuts(n)) {
+                worklist.enqueue(out.node);
+            }
+        }
+
+        private Node[] iterableToArray(Iterable<Node> iterable) {
+            List<Node> list = new ArrayList<>();
+            for(var entry : iterable) {
+                list.add(entry);
+            }
+            return list.toArray(new Node[0]);
+        }
+
+        @Override
+        public void visit(Add node) {
+            visitPreds(node, iterableToArray(node.getPreds()), binding_irnode.ir_opcode.iro_Add);
+        }
+
+        @Override
+        public void visit(And node) {
+            visitPreds(node, iterableToArray(node.getPreds()), binding_irnode.ir_opcode.iro_And);
+        }
+
+        @Override
+        public void visit(Cmp node) {
+            // do nothing
+        }
+
+        @Override
+        public void visit(Cond node) {
+            // do nothing
+        }
+
+        @Override
+        public void visit(Proj proj) {
+            // do nothing
+        }
+
+        @Override
+        public void visit(Call call) {
+            visitPreds(call, iterableToArray(call.getPreds()), binding_irnode.ir_opcode.iro_Call);
+        }
+
+        @Override
+        public void visit(Const node) {
```

Comment:
If you want to eliminate function calls, you also need to handle `Address` nodes in some way. They should be pretty much like `Const`, just using the referenced `Entitiy` as key, instead of the `TargetValue`.